integration_tests/common.rs

```diff
@@ -0,0 +1,261 @@
+    let merchant_config = merchant_test_config().await;
+
+    // set up tracing for all of our own log messages
+    tracing_subscriber::fmt()
+        .with_writer(Mutex::new(
+            File::create(ERROR_FILENAME).expect("Failed to open log file"),
+        ))
+        .init();
+
+    // Form the merchant run request and execute
+    let run = merchant_cli!(Run, vec!["run"]);
+    let merchant_handle = tokio::spawn(
+        run.run(merchant_config)
+            .instrument(info_span!(Party::MerchantServer.to_str())),
+    );
+
+    // Form the customer watch request and execute
+    let watch = customer_cli!(Watch, vec!["watch"]);
+    let customer_handle = tokio::spawn(
+        watch
+            .run(rng.clone(), customer_config)
+            .instrument(info_span!(Party::CustomerWatcher.to_str())),
+    );
+
+    future::join(customer_handle, merchant_handle)
+}
+
+pub async fn teardown(server_future: ServerFuture) {
+    // Ignore the result because we expect it to be an `Expired` error
+    let _result = server_future
+        .with_timeout(tokio::time::Duration::new(1, 0))
+        .await;
+
+    // delete data from this run
+    let _ = fs::remove_dir_all("integration_tests/gen/");
+}
+
+/// Encode the customizable fields of the zeekoe customer Config struct for testing.
+async fn customer_test_config() -> zeekoe::customer::Config {
+    let m = HashMap::from([
+        ("database", "{ sqlite = \"customer.db\" }"),
+        ("trust_certificate", "\"localhost.crt\""),
+        ("tezos_account", "{ alias = \"alice\" }"),
+        ("tezos_uri", "\"http://localhost:20000\""),
+        ("self_delay", "120"),
+        ("confirmation_depth", "1"),
+    ]);
+
+    let contents = m.into_iter().fold("".to_string(), |acc, (key, value)| {
+        format!("{}{} = {}\n", acc, key.to_string(), value.to_string())
+    });
+
+    write_config_file(CUSTOMER_CONFIG, contents);
+
+    zeekoe::customer::Config::load(CUSTOMER_CONFIG)
+        .await
+        .expect("Failed to load customer config")
+}
+
+/// Encode the customizable fields of the zeekoe merchant Config struct for testing.
+async fn merchant_test_config() -> zeekoe::merchant::Config {
+    let m = HashMap::from([
+        ("database", "{ sqlite = \"merchant.db\" }"),
+        ("tezos_account", "{ alias = \"bob\" }"),
+        ("tezos_uri", "\"http://localhost:20000\""),
+        ("self_delay", "120"),
+        ("confirmation_depth", "1"),
+    ]);
+
+    let tezos_contents = m.into_iter().fold("".to_string(), |acc, (key, value)| {
+        format!("{}{} = {}\n", acc, key.to_string(), value.to_string())
+    });
+
+    // helper to write out the service for ipv4 and v6 localhost addresses
+    let generate_service = |addr: &str| {
+        HashMap::from([
+            ("address", addr),
+            ("private_key", "localhost.key"),
+            ("certificate", "localhost.crt"),
+        ])
+        .into_iter()
+        .fold("\n[[service]]".to_string(), |acc, (key, value)| {
```

Comment:
Isn't this printing [[service]] between every line?

Reply:
Nope, the first argument to `fold` is the initial value for the accumulator (`acc`). So the first time the second argument function is called, its arguments are `("[[service]]", ("address", "localhost"))`. Then the next time, the arguments are `("[[service]]\n address = localhost", ("private_key", "localhost.key")` and so on.